---
title: "Introduction to Solving Biological Problems Using R - Day 2"
author: Mark Dunning, Suraj Menon and Aiora Zabala. Original material by Robert Stojnić,
  Laurent Gatto, Rob Foy, John Davey, Dávid Molnár and Ian Roberts
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_notebook
---


# 3. Data Manipulation Techniques

## Motivation

- So far we have been lucky that all our data have been in the same file:
    + This is not usually the case
    + Dataset may be spread over several files
        + This takes longer, and is harder, than many people realise
    + We need to combine before doing an analysis



## Combining data from multiple sources: Gene Clustering Example

- R has powerful functions to combine heterogeneous data sources into a single data set
- Gene clustering example data:
    + Gene expression values in ***gene.expression.txt***
    + Gene information in ***gene.description.txt***
    + Patient information in ***cancer.patients.txt***
- A breast cancer dataset with numerous patient characteristics:
    + We will concentrate on ***ER status*** (positive / negative)
    + What genes show a statistically-significant different change between ER groups?

## Analysis goals

- We will show how to lookup a particular gene in the dataset
- Also, how to look-up genes in a given genomic region
- Assess if a given gene is differentially-expressed
- Create a heatmap to cluster the samples and reveal any subgroups in the data

## Peek at the data

```{r}
evals <- read.delim("gene.expression.txt")
evals[1:2,1:5]
```


- `r nrow(evals)` rows and `r ncol(evals)` columns
+ One row for each gene:
    + Rows are named according to particular technology used to make measurement
    + The names of each row can be returned by `rownames(evals)`; giving a vector
+ One column for each patient:
    + The names of each column can be returned by `colnames(evals)`; giving a vector
    

```{r}
genes <- read.delim("gene.description.txt",stringsAsFactors = FALSE)
head(genes)
```


- `r nrow(genes)` rows and `r ncol(genes)` columns
- One for each gene
- Includes mapping between manufacturer ID and Gene name


```{r}
subjects <- read.delim("cancer.patients.txt",stringsAsFactors = FALSE)
head(subjects)
```

- One for each patient in the study
- Each column is a different characteristic of that patient
    + e.g. whether a patient is ER positive (value of 1) or negative (value of 0)

```{r}
table(subjects$er)
```



## Ordering and sorting

To get a feel for these data, we will look at how we can subset and order

- R allows us to do the kinds of filtering, sorting and ordering operations you might be familiar with in Excel
- For example, if we want to get information about patients that are ER negative
    + these are indicated by an entry of ***0*** in the `er` column

```{r}
subjects$er == 0
```

We can do the comparison within the square brackets

- Remembering to include a `,` to index the columns as well
- Best practice to create a new variable and leave the original data frame untouched

```{r}
erNegPatients <- subjects[subjects$er == 0,]
head(erNegPatients)
```

or

```{r}
View(erNegPatients)
```

Sorting is supported by the **`sort()`** function

- Given a vector, it will return a sorted version of the same length

```{r}
sort(erNegPatients$grade)
```

- But this is not useful in all cases
    + We have lost the extra information that we have about the patients
    
- Instead, we can use **`order()`**
- Given a vector, `order()` will give a set of numeric values which will give an ordered version of the vector
    + default is smallest --> largest

```{r}
myvec <- c(90,100,40,30,80,50,60,20,10,70)
myvec
order(myvec)
```

- i.e. number in position 9 is the smallest, number in position 8 is the second smallest:

```{r}
myvec[9]
myvec[8]
```

N.B. `order` will also work on character vectors 

```{r}
firstName  <- c("Adam", "Eve", "John", "Mary", "Peter", "Paul", "Joanna", "Matthew", "David", "Sally")
order(firstName)
```

- We can use the result of `order()` to perform a subset of our original vector
- The result is an ordered vector
```{r}
myvec.ord <- myvec[order(myvec)]
myvec.ord
```

- Implication: We can use `order` on a particular column of a data frame, and use the result to sort all the rows

- We might want to select the youngest ER negative patients for a follow-up study
- Here we order the `age` column and use the result to re-order the rows in the data frame

```{r}
erNegPatientsByAge <- erNegPatients[order(erNegPatients$age),]
head(erNegPatientsByAge)
```


- can change the behaviour of `order` to be Largest --> Smallest
```{r}
erNegPatientsByAge <- erNegPatients[order(erNegPatients$age,decreasing = TRUE),]
head(erNegPatientsByAge)
```

- we can write the result to a file if we wish

```{r eval=FALSE}
write.table(erNegPatientsByAge, file="erNegativeSubjectsByAge.txt", sep="\t")
```


## Exercise: Exercise7 

- Imagine we want to know information about chromosome 8 genes that have been measured.
1. Create a new data frame containing information on genes on Chromosome 8
2. Order the rows in this data frame according to start position, and write the results to a file

```{r}

## Your Answer Here ###

```


## Alternative: 

- you might find the function `subset` a bit easier to use
    + no messing around with square brackets
    + no need to remember row and column indices
    + no need for `$` operator to access columns
- more advanced packages like dplyr use a similar approach
    + you'll find out about this on our intermediate course
    
```{r}
chr8Genes <- subset(genes, Chromosome=="chr8")
head(chr8Genes)
```


## Retrieving data for a particular gene

 - Gene `ESR1` is known to be hugely-different between ER positive and negative patient
    + let's check that this is evident in our dataset
    + if not, something has gone wrong!
- First step is to locate this gene in our dataset
- We can use `==` to do this, but there are some alternatives that are worth knowing about

## Character matching in R

- `match()` and `grep()` are often used to find particular matches
    + CAUTION: by default, match will only return the ***first*** match!

```{r}
match("D", LETTERS)
grep("F", rep(LETTERS,2))
match("F", rep(LETTERS,2))
```

- `grep` can also do partial matching
    + can also do complex matching using "regular expressions"
    
```{r}
month.name
grep("ary",month.name)
grep("ber",month.name)
```

- `%in%` will return a logical if each element is contained in a shortened list

```{r}
month.name %in% c("May", "June")
```


## Retrieving data for a particular gene

- Find the name of the ID that corresponds to gene ***ESR1*** using `match`
    + mapping between IDs and genes is in the ***genes*** data frame
        + ID in first column, gene name in the second
- Save this ID as a variable

```{r}
ind <- match("ESR1", genes$HUGO.gene.symbol)
genes[ind,]
probe <- genes[ind,1]
probe
```

Now, find which row in our expression matrix is indexed by this ID

- recall that the rownames of the expression matrix are the probe IDs
- save the expression values as a variable

```{r}
match(probe, rownames(evals))
evals[match(probe, rownames(evals)), 1:10]
genevals <- evals[match(probe,rownames(evals)),]
class(genevals)
```


    
## Relating to patient characteristics

We have expression values and want to visualise them against our categorical data

- use a boxplot, for example
- however, we have to first make sure our values are treat as numeric data
- as we created the subset of a data frame, the result was also a data frame
    + use `as.numeric`


```{r}
boxplot(as.numeric(genevals) ~ subjects$er)
```


- In this case there is a clear difference, so we probably don't even need a p-value to convince ourselves of the difference
    + in real-life, we would probably test lots of genes and implement some kind of multiple-testing
    + e.g. `p.adjust` (`?p.adjust`)

```{r}
t.test(as.numeric(genevals) ~ subjects$er)

```



## Complete script

```{r}
genes    <- read.delim("gene.description.txt",stringsAsFactors = FALSE)
subjects <- read.delim("cancer.patients.txt",stringsAsFactors = FALSE)
evals    <- read.delim("gene.expression.txt")

ind      <- match("ESR1", genes[,2])
probe    <- genes[ind,1]
genevals <- evals[match(probe,rownames(evals)),]

boxplot(as.numeric(genevals) ~ subjects$er)
t.test(as.numeric(genevals) ~ subjects$er)
```

## Exercise: Exercise 8

Repeat the same steps we performed for the gene ESR1, but for GATA3:

- Try and make as few changes as possible from the ESR1 script
- Can you see why making a markdown document is useful for analysis?


```{r}

### Your Answer Here ###

```



